python/paddle/fluid/trainer.py

```diff
@@ -30,6 +33,48 @@ def __init__(self):
         self.type = Event.BEGIN_EPOCH
 
 
+def dist_traspile_if_necessary(startup_prog, main_prog, optimize_ops,
```

Comment:
I think the user should not need to change code when switching from local training to distributed training. E.g., the user should not need to switch from `fluid.Trainer` to `fluid.DistributedTrainer`.

Still, we can use the same external API (e.g., `fluid.Trainer`), and internally have a class hierarchy for code reuse. Since currently we are not sure what the code would look like, I would prefer to refactor it if needed.